modules/renter/contractor/contracts.go

```diff
@@ -0,0 +1,388 @@
+	// period.
+	c.mu.RLock()
+	endHeight := c.blockHeight + c.allowance.Period
+	max, err := maxSectors(c.allowance, c.hdb, c.tpool)
+	c.mu.RUnlock()
+	if err != nil {
+		return
+	}
+	// Only allocate half as many sectors as the max. This leaves some leeway
+	// for replacing contracts, transaction fees, etc.
+	numSectors := max / 2
+	// check that this is sufficient to store at least one sector
+	if numSectors == 0 {
+		return
+	}
+
+	// Loop through the contracts and renew them one-by-one.
+	for _, id := range renewSet {
+		// Renew one contract.
+		func() {
+			// Mark the contract as being renewed, and defer logic to unmark it
+			// once renewing is complete.
+			c.mu.Lock()
+			c.renewing[id] = true
+			c.mu.Unlock()
+			defer func() {
+				c.mu.Lock()
+				delete(c.renewing, id)
+				c.mu.Unlock()
+			}()
+
+			// Wait for any active editors and downloaders to finish for this
+			// contract, and then grab the latest revision.
+			c.mu.RLock()
+			e, eok := c.editors[id]
+			d, dok := c.downloaders[id]
+			c.mu.RUnlock()
+			if eok {
+				e.invalidate()
+			}
+			if dok {
+				d.invalidate()
+			}
+
+			c.mu.RLock()
+			oldContract, ok := c.contracts[id]
+			c.mu.RUnlock()
+			if !ok {
+				c.log.Println("WARN: no record of contract previously added to the renew set:", id)
+				return
+			}
+
+			// Create the new contract.
+			newContract, err := c.managedRenew(oldContract, numSectors, endHeight)
+			if err != nil {
+				c.log.Printf("WARN: failed to renew contract %v with %v: %v\n", id, oldContract.NetAddress, err)
+				return
+			}
+			c.log.Printf("Renewed contract %v with %v\n", id, oldContract.NetAddress)
+
+			// Lock the contractor as we update it to use the new contract
+			// instead of the old contract.
+			c.mu.Lock()
+			defer c.mu.Unlock()
+
+			// Store the contract in the record of historic contracts.
+			_, exists := c.contracts[oldContract.ID]
+			if exists {
+				c.oldContracts[oldContract.ID] = oldContract
+				delete(c.contracts, oldContract.ID)
+			}
+
+			// Add the new contract, including a mapping from the old
+			// contract to the new contract.
+			c.contracts[newContract.ID] = newContract
+			c.renewedIDs[oldContract.ID] = newContract.ID
+			c.cachedRevisions[newContract.ID] = c.cachedRevisions[oldContract.ID]
+			delete(c.cachedRevisions, oldContract.ID)
+
+			// Save the contractor.
+			err = c.saveSync()
+			if err != nil {
+				c.log.Println("Failed to save the contractor after creating a new contract.")
+			}
+		}()
+
+		// Soft sleep for a minute to allow all of the transactions to propagate
+		// the network.
+		select {
+		case <-c.tg.StopChan():
+			return
+		case <-time.After(contractFormationInterval):
+		}
+	}
+
+	// Quit in the event of shutdown.
+	select {
+	case <-c.tg.StopChan():
+		return
+	default:
+	}
+
+	// Count the number of contracts, and add more where they are needed.
+	//
+	// TODO: Use the IsGoodForUpload logic.
```

Comment:
cool, i'll rebase to this pr and then use this new flag
now the 1 sector contract strategy seems not work